packages/utils/src/diff/index.ts

```diff
@@ -336,13 +336,60 @@ export function replaceAsymmetricMatcher(
     const expectedValue = expected[key]
     const actualValue = actual[key]
     if (isAsymmetricMatcher(expectedValue)) {
-      if (expectedValue.asymmetricMatch(actualValue)) {
-        actual[key] = expectedValue
+      const matches = expectedValue.asymmetricMatch(actualValue)
+      // For container matchers (ArrayContaining, ObjectContaining), unwrap and recursively process
+      if ('sample' in expectedValue && expectedValue.sample !== undefined && isReplaceable(actualValue, expectedValue.sample)) {
+        if (matches) {
+          // Matcher matches: unwrap and recursively process to show actual structure
+          const replaced = replaceAsymmetricMatcher(
+            actualValue,
+            expectedValue.sample,
+            actualReplaced,
+            expectedReplaced,
+          )
+          actual[key] = replaced.replacedActual
+          expected[key] = replaced.replacedExpected
```

Comment:
Testing hypothesis of https://github.com/vitest-dev/vitest/pull/9330/files#r2645080058

```suggestion
          expected[key] = actualValue
```